### ISSUE TITLE
请教下关于DAN的transfer_loss计算

### ISSUE THREAD

**Issue by jizefeng0810** [jindongwang/transferlearning#124]
**Describe the bug**
王博对DAN的多核stransfer_loss计算如下
`def mmd_rbf_noaccelerate(source, target, kernel_mul=2.0, kernel_num=5, fix_sigma=None):
    batch_size = int(source.size()[0])
    kernels = guassian_kernel(source, target,
                              kernel_mul=kernel_mul, kernel_num=kernel_num, fix_sigma=fix_sigma)
    XX = kernels[:batch_size, :batch_size]
    YY = kernels[batch_size:, batch_size:]
    XY = kernels[:batch_size, batch_size:]
    YX = kernels[batch_size:, :batch_size]
    loss = torch.mean(XX + YY - XY -YX)
    return loss`
`def mmd_rbf_accelerate(source, target, kernel_mul=2.0, kernel_num=5, fix_sigma=None):
    batch_size = int(source.size()[0])
    kernels = guassian_kernel(source, target,
        kernel_mul=kernel_mul, kernel_num=kernel_num, fix_sigma=fix_sigma)
    loss = 0
    for i in range(batch_size):
        s1, s2 = i, (i+1)%batch_size
        t1, t2 = s1+batch_size, s2+batch_size
        loss += kernels[s1, s2] + kernels[t1, t2]
        loss -= kernels[s1, t2] + kernels[s2, t1]
    return loss / float(batch_size)`

**To Reproduce**
为什么我使用
`mmd_rbf_noaccelerate`损失数值大于零
而在mmd_rbf_accelerate损失数值小于零？按照paper的定义，这里的损失应该是distance，不应该出现负数，而上面的函数大于零，是王博在龙明盛老师提供的代码上修改的，想请问为什么如此计算，龙明盛老师的这段代码为什么计算结果是负数？理论上distance不应该一个小于0的数

`lambd = 2 / (1 + math.exp(-10 * (i) / iteration)) - 1   
        loss = cls_loss + lambd * mmd_loss `
另请问王博为什么这样计算lambd的值？原代码是使用lambd=1
望王博解答，谢谢！
**Screenshots**
下面是mmd_rbf_noaccelerate输出结果，最后为transfer_loss
Train iter: 10 [(0%)]	Loss: 4.239232	soft_Loss: 3.447802	mmd_Loss: 0.791430
下面是mmd_rbf_accelerate输出结果，最后为transfer_loss
Train iter: 10 [(0%)]	Loss: 3.372070	soft_Loss: 3.522608	mmd_Loss: -0.150538

**Comment by jindongwang** [comment]
那份DAN代码我后来重新整理了，你看DDC_CORAL的代码吧，一样的意思。

**Comment by jizefeng0810** [comment]
王博，我看了DDC_CORAL的代码，发现transfer_loss部分和 mmd_rbf_noaccelerate()是一样的，计算后损失是正值，但是用龙明盛老师的代码算了是负值，为什么呢？还有transfer_loss的损失一直不降，是正常的吗？DAN中只使用了一层计算距离，这和paper里面的不一样，没有进行多层的适配。

**Comment by jindongwang** [comment]
龙老师放出来的DAN pytorch代码就是适配了一层，之前我和他的代码的作者说过，他表示就是这样的。可能他们原始的caffe模型是三层吧，这个我就没在管了。从这个意义上说，DAN就是MK-MMD版本的DDC。所以我现在主要在修改DDC模型了。那个DAN的code比较旧了，不建议你看。
损失的部分在我电脑上是可以降低的，没出现过负数。我建议你不要纠结这个代码对比，你按照你的理解看一下代码，如果觉得没问题就可以用

**Comment by jizefeng0810** [comment]
王博我用了你的代码，总损失是降的，transfer_loss是上升的，这是正常的吗




| |
纪泽锋
|
|
邮箱：jizefeng0810@163.com
|

签名由 网易邮箱大师 定制

在2019年11月04日 16:31，Jindong Wang 写道：

龙老师放出来的DAN pytorch代码就是适配了一层，之前我和他的代码的作者说过，他表示就是这样的。可能他们原始的caffe模型是三层吧，这个我就没在管了。从这个意义上说，DAN就是MK-MMD版本的DDC。所以我现在主要在修改DDC模型了。那个DAN的code比较旧了，不建议你看。
损失的部分在我电脑上是可以降低的，没出现过负数。我建议你不要纠结这个代码对比，你按照你的理解看一下代码，如果觉得没问题就可以用

**Comment by jindongwang** [comment]
那你把loss vs. epoch的图画出来我看一下是不是和我的一样

**Comment by jizefeng0810** [comment]
![image](https://user-images.githubusercontent.com/35838455/68127514-03a3a180-ff51-11e9-8403-9aee8a6d963c.png)
王博，这是源域数据是webcam，目标域数据是amazon的损失曲线，这是我改了一点网络结构的

**Comment by jindongwang** [comment]
1、你改了网络结构，是怎么改的，会不会有影响？
2、你这个图的横坐标是什么？

**Comment by jizefeng0810** [comment]
我只是在resnet后面直接加了两层FC。和你的代码其实差不多。
横坐标是总的loss计算次数，我把每次都记录下来了




| |
纪泽锋
|
|
邮箱：jizefeng0810@163.com
|

签名由 网易邮箱大师 定制

在2019年11月05日 10:13，Jindong Wang 写道：

1、你改了网络结构，是怎么改的，会不会有影响？
2、你这个图的横坐标是什么？

**Comment by jindongwang** [comment]
我帮你跑了一下，你可以看看代码中的train_log.csv，与main对应一下看每个epoch的loss。总的来说loss都是下降趋势的，可能mmd的loss下降不是特别大。你为什么要记录每次loss计算呢？我们都是看每个epoch整体的loss。我觉得你可以好好学习一下代码再问。

**Comment by jindongwang** [comment]
@jizefeng0810 请问你的问题解决了吗

**Comment by jizefeng0810** [comment]
解决了，谢谢王博




| |
纪泽锋
|
|
邮箱：jizefeng0810@163.com
|

签名由 网易邮箱大师 定制

在2019年11月15日 17:48，Jindong Wang 写道：

@jizefeng0810 请问你的问题解决了吗

**Comment by jindongwang** [comment]
怎么解决的呢？

**Comment by jizefeng0810** [comment]
km-mmd损失没有降，但是效果确实比不用mmd好一些，我只能理解为km-mmd损失尽可能减缓了源域和目标域特征分布的发散


| |
纪泽锋
|
|
邮箱：jizefeng0810@163.com
|

签名由 网易邮箱大师 定制

在2019年11月15日 17:50，Jindong Wang 写道：

怎么解决的呢？
